plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,251 @@
+#include "irods/switch_user.h"
+#include "irods/irods_rs_comm_query.hpp"
+#include "irods/rodsErrorTable.h"
+#include "irods/irods_logger.hpp"
+#include "irods/server_utilities.hpp"
+#include "irods/version.hpp"
+
+#define IRODS_USER_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+#include "irods/user_administration.hpp"
+
+#include <cstring>
+#include <string>
+#include <string_view>
+// clang-format on
+
+// This global variable is required so that the API plugin can access RcComm connections
+// created as a result of previous API calls (i.e. connections created due to redirection).
+extern zoneInfo* ZoneInfoHead;
+
+namespace
+{
+    //
+    // Namespace and Type aliases
+    //
+
+    using log_api = irods::experimental::log::api;
+
+    //
+    // Function Prototypes
+    //
+
+    auto call_switch_user(irods::api_entry*, RsComm*, SwitchUserInput*) -> int;
+    auto rs_switch_user(RsComm*, SwitchUserInput*) -> int;
+    auto check_input(const SwitchUserInput*) -> int;
+    auto update_server_to_server_connections(RsComm&) -> void;
+
+    //
+    // Function Implementations
+    //
+
+    auto call_switch_user(irods::api_entry* _api, RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        return _api->call_handler<SwitchUserInput*>(_comm, _input);
+    } // call_switch_user
+
+    auto check_input(const SwitchUserInput* _input) -> int
+    {
+        if (!_input) {
+            log_api::error("Invalid input argument: received a null pointer");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strnlen(_input->username) == 0) {
+            log_api::error("Invalid username argument: null or empty");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strlen(_input->zone) == 0) {
+            log_api::error("Invalid zone argument: null or empty");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        return 0;
+    } // check_input
+
+    auto update_server_to_server_connections(RsComm& _comm) -> void
+    {
+        const auto switch_user_or_disconnect = [&client = _comm.clientUser](rodsServerHost& _host) {
+            if (!_host.conn) {
+                log_api::warn("No connection to remote host."); // TODO Consider including the host's name.
+                return;
+            }
+
+            // Disconnect if the remote server's version is older than 4.3.1.
+            // Remember, only iRODS 4.3.1 and later support the switch user API plugin.
+            if (*irods::to_version(_host.conn->svrVersion->relVersion) < irods::version{4, 3, 1}) {
+                rcDisconnect(_host.conn);
+                return;
+            }
+
+            // At this point, we know the remote server supports this API plugin.
+            if (const auto ec = rc_switch_user(_host.conn, client.userName, client.rodsZone); ec != 0) {
+                log_api::error(
+                    "rc_switch_user failed on remote server [error_code=[{}]]. Disconnecting from remote server.", ec);
```

Comment:
same here?   helpful to write remote server's hostname if we already have it?

Reply:
done